src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,688 @@
+            {
+                return null;
+            }
+
+            var actions = AddActions(semanticDocument, expression, cancellationToken);
+
+            if (actions.Length > 0)
+            {
+                return new CodeActionWithNestedActions(FeaturesResources.Introduce_parameter, actions, isInlinable: true);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Introduces a new parameter and refactors all the call sites
+        /// </summary>
+        public async Task<Solution> IntroduceParameterAsync(SemanticDocument document, TExpressionSyntax expression,
+            bool allOccurrences, bool trampoline, bool overload, CancellationToken cancellationToken)
+        {
+            var parameterName = GetNewParameterName(document, expression, cancellationToken);
+
+            // MethodSymbol not null here since we know we're contained in something containing parameters at this point
+            var methodSymbolInfo = (IMethodSymbol)document.SemanticModel.GetRequiredEnclosingSymbol(expression.SpanStart, cancellationToken);
+            var methodCallSites = await FindCallSitesAsync(document, methodSymbolInfo, cancellationToken).ConfigureAwait(false);
+
+            if (trampoline || overload)
+            {
+                return await RewriteSolutionWithNewMethodAsync(document,
+                expression, methodSymbolInfo, allOccurrences, parameterName, methodCallSites, trampoline, cancellationToken).ConfigureAwait(false);
+            }
+
+            return await RewriteSolutionAsync(document,
+                expression, methodCallSites, allOccurrences, parameterName, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected static string GetNewParameterName(SemanticDocument document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            if (ShouldRemoveVariableDeclaratorContainingExpression(document, expression, out var varDecl))
+            {
+                return varDecl;
+            }
+
+            var semanticFacts = document.Document.GetRequiredLanguageService<ISemanticFactsService>();
+            return semanticFacts.GenerateNameForExpression(document.SemanticModel, expression, capitalize: false, cancellationToken);
+        }
+
+        protected static bool ShouldRemoveVariableDeclaratorContainingExpression(
+            SemanticDocument document, TExpressionSyntax expression, out string varDecl)
+        {
+            var syntaxFacts = document.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var expressionDecl = expression.GetRequiredParent().GetRequiredParent();
+            varDecl = "";
+
+            if (syntaxFacts.IsVariableDeclarator(expressionDecl))
+            {
+                varDecl = syntaxFacts.GetIdentifierOfVariableDeclarator(expressionDecl).ValueText;
+                return true;
+            }
+
+            return false;
```

Comment:
```suggestion
            varDecl = null;
```

Reply:
still needs to return false post setting it to null

Reply:
yes.  this is the right pattern to use :)